File: code/cvc5mvn/src/main/java/plaid/antlr/CommandVisitor.java
package plaid.antlr;

import plaid.PreludeBaseVisitor;
import plaid.PreludeParser.AssertCommandContext;
import plaid.PreludeParser.AssignCommandContext;
import plaid.PreludeParser.CommandListContext;
import plaid.PreludeParser.FunctionCallCommandContext;
import plaid.PreludeParser.LetCommandContext;
import plaid.ast.AssertCmd;
import plaid.ast.AssignCmd;
import plaid.ast.ListCmd;
import plaid.ast.CallCmd;
import plaid.ast.Identifier;
import plaid.ast.LetCmd;
import plaid.ast.Cmd;

public class CommandVisitor extends PreludeBaseVisitor<Cmd> {

    @Override
    public CallCmd visitFunctionCallCommand(FunctionCallCommandContext ctx) {
        return new CallCmd(
                new Identifier(ctx.ident().getText()),
                ctx.expr().stream().map(Loader::toExpression).toList());
    }

    @Override
    public AssignCmd visitAssignCommand(AssignCommandContext ctx) {
        return new AssignCmd(
                Loader.toExpression(ctx.expr(0)),
                Loader.toExpression(ctx.expr(1)));
    }

    @Override
    public AssertCmd visitAssertCommand(AssertCommandContext ctx) {
        return new AssertCmd(
                Loader.toExpression(ctx.expr(0)),
                Loader.toExpression(ctx.expr(1)),
                Loader.toExpression(ctx.expr(2)));
    }

    @Override
    public ListCmd visitCommandList(CommandListContext ctx) {
        return new ListCmd(visit(ctx.command(0)), visit(ctx.command(1)));
    }

    @Override
    public LetCmd visitLetCommand(LetCommandContext ctx) {
        return new LetCmd(
                new Identifier(ctx.ident().getText()),
                Loader.toExpression(ctx.expr()),
                Loader.toCommand(ctx.command()));
    }

}

File: code/cvc5mvn/src/main/scala/plaid/antlr/CommandVisitor.scala
package plaid.antlr

import plaid.PreludeBaseVisitor
import plaid.PreludeParser.*
import plaid.ast.*

import scala.jdk.CollectionConverters.*

object CommandVisitor extends PreludeBaseVisitor[Cmd] {

  override def visitFunctionCallCommand(ctx: FunctionCallCommandContext): CallCmd =
    CallCmd(
      Identifier(ctx.ident().getText),
      ctx.expr().asScala.map(Load.expression).toList
    )

  override def visitAssignCommand(ctx: AssignCommandContext): AssignCmd =
    AssignCmd(
      Load.expression(ctx.expr(0)),
      Load.expression(ctx.expr(1))
    )

  override def visitAssertCommand(ctx: AssertCommandContext): AssertCmd =
    AssertCmd(
      Load.expression(ctx.expr(0)),
      Load.expression(ctx.expr(1)),
      Load.expression(ctx.expr(2))
    )

  override def visitCommandList(ctx: CommandListContext): ListCmd =
    ListCmd(
      visit(ctx.command(0)),
      visit(ctx.command(1))
    )

  override def visitLetCommand(ctx: LetCommandContext): LetCmd =
    LetCmd(
      Identifier(ctx.ident().getText),
      Load.expression(ctx.expr()),
      Load.command(ctx.command())
    )
}
